src/memory_manager.rs

```diff
@@ -799,8 +821,144 @@ fn bucket_allocations_address(id: BucketId) -> Address {
+            );
+        }
+
+        BucketId(u16::from_be_bytes(next_bits.data))
+    }
+
+    // Sets the next bucket in the linked list of buckets
+    fn set_next(&mut self, bucket: BucketId, next: BucketId) {
+        let start_bit_index = bucket.0 as usize * 15;
+        let next_bits: BitArray<[u8; 2]> = BitArray::from(next.0.to_be_bytes());
+
+        for (index, bit) in next_bits.iter().skip(1).enumerate() {
+            self.inner
+                .bucket_links
+                .set(start_bit_index + index, bit.as_bool());
+        }
+
+        let start_byte_index = start_bit_index / 8;
+        let end_byte_index = (start_bit_index + 14) / 8;
+
+        self.dirty_bucket_link_bytes
+            .extend((start_byte_index..=end_byte_index).map(|i| i as u16))
+    }
+
+    // Calculates the buckets for a given memory by iterating over its linked list
+    fn buckets_for_memory(&self, memory_id: MemoryId, count: u16) -> Vec<BucketId> {
+        if count == 0 {
+            return Vec::new();
+        }
+
+        let mut bucket = self.get_first(memory_id);
+        let mut buckets = vec![bucket];
+        for _ in 1..count {
+            bucket = self.get_next(bucket);
+            buckets.push(bucket);
+        }
+        buckets
+    }
+
+    // Flushes only the dirty bytes to memory
+    fn flush_dirty_bytes<M: Memory>(&mut self, memory: &M, start_offset: u64) {
+        const FIRST_BUCKET_PER_MEMORY_LEN: usize = 2 * MAX_NUM_MEMORIES as usize;
+
+        if !self.dirty_first_buckets.is_empty() {
+            // SAFETY: This is safe because we simply cast from [u16] to [u8] and double the length.
+            let bytes: [u8; FIRST_BUCKET_PER_MEMORY_LEN] =
+                unsafe { transmute(self.inner.first_bucket_per_memory) };
+
+            // Multiply by 2 since we've converted from [u16] to [u8].
+            let min = 2 * self.dirty_first_buckets.first().unwrap().0 as usize;
+            let max = 2 * self.dirty_first_buckets.last().unwrap().0 as usize + 1;
+
+            let slice = &bytes[min..=max];
+            write(memory, start_offset + min as u64, slice);
+            self.dirty_first_buckets.clear();
+        }
+
+        if !self.dirty_bucket_link_bytes.is_empty() {
+            let min = *self.dirty_bucket_link_bytes.first().unwrap() as usize;
+            let max = *self.dirty_bucket_link_bytes.last().unwrap() as usize;
+
+            let slice = &self.inner.bucket_links.data[min..=max];
+            write(
+                memory,
+                start_offset + (FIRST_BUCKET_PER_MEMORY_LEN + min) as u64,
+                slice,
+            );
+            self.dirty_bucket_link_bytes.clear();
+        }
+    }
+
+    // Flushes all bytes to memory
+    fn flush_all<M: Memory>(&mut self, memory: &M, start_offset: u64) {
+        write_struct(&self.inner, Address::from(start_offset), memory);
+        self.dirty_first_buckets.clear();
+        self.dirty_bucket_link_bytes.clear();
+    }
+}
```

Comment:
I would strongly prefer to have only one place where we read the header, and after that, we use u16 (instead of 15-bit representation in storage). I believe that will improve performance, but more importantly, it will be easier to maintain and understand. After that, we can still use information from dirty_first_buckets and dirty_bucket_link_bytes to apply changes in storage on 15-bit representation when saving.

Reply:
Yeah that makes sense.
This means we'll have to convert from 15 to 16-bit after reading which will have some overhead but at least its only done once. I'll have a go at it now.

Reply:
Done. Will likely need a bit of cleaning up. Need to run off now but will give it another look tomorrow.